File: src/main/java/org/flossware/civilization/module/TechnologyModule.java
package org.flossware.civilization.module;

import org.flossware.civilization.engine.TickContext;
import org.flossware.civilization.model.*;
import org.flossware.civilization.model.Event.EventSeverity;
import org.flossware.civilization.model.Event.EventType;
import org.flossware.civilization.util.SeedManager;

import java.util.*;

/**
 * Technology research and diffusion module.
 * Pure function with explicit DAG-based unlocking.
 */
public final class TechnologyModule implements SimulationModule {

    private static final double BASE_RESEARCH_POINTS = 10.0;
    private static final double UNIVERSITY_BONUS = 0.5;
    private static final double MAX_POPULATION_FACTOR = 1.0;
    private static final double POPULATION_SCALING_DIVISOR = 1_000_000.0;
    private static final double TRADE_CONNECTIVITY_SCALE = 0.2;

    private final TechGraph techGraph;

    /**
     * Creates a TechnologyModule with the given tech graph for DAG-based unlocking.
     *
     * @param techGraph the technology dependency graph
     */
    public TechnologyModule(TechGraph techGraph) {
        this.techGraph = Objects.requireNonNull(techGraph);
    }

    @Override
    public String moduleName() {
        return "technology";
    }

    @Override
    public ModuleResult<?> tick(TickContext context) {
        var random = SeedManager.getModuleRandom(context.yearSeed(), moduleName());
        double tradeConnectivity = Math.min(1.0,
            context.state().economy().tradeRoutes().size() * TRADE_CONNECTIVITY_SCALE);
        return tick(
            context.state().technology(),
            techGraph,
            tradeConnectivity,
            context.state().population().population(),
            random
        );
    }

    @SuppressWarnings("unchecked")
    @Override
    public CivilizationState applyResult(CivilizationState state, ModuleResult<?> result) {
        return state.withTechnology(((ModuleResult<TechnologyState>) result).state());
    }

    public static ModuleResult<TechnologyState> tick(
        TechnologyState current,
        TechGraph techGraph,
        double tradeConnectivity,
        long population,
        SplittableRandom random
    ) {
        List<Event> events = new ArrayList<>();
        TechnologyState newState = current;

        // Calculate research points per year
        double researchPoints = calculateResearchPoints(current, population);

        // Find next researchable technology
        Optional<Technology> nextTech = findNextResearchableTech(current, techGraph);

        if (nextTech.isPresent()) {
            Technology tech = nextTech.get();
            double currentProgress = current.researchProgress().getOrDefault(tech.id(), 0.0);
            double newProgress = currentProgress + researchPoints;

            if (newProgress >= tech.researchCost()) {
                // Unlock technology
                newState = newState.withUnlockedTech(tech.id());
                events.add(new Event(0, "", EventType.TECHNOLOGY_UNLOCKED, EventSeverity.MAJOR,
                    "Unlocked: " + tech.id(), tech));

                // Clear progress
                Map<String, Double> updatedProgress = new HashMap<>(current.researchProgress());
                updatedProgress.remove(tech.id());
                newState = new TechnologyState(newState.unlockedTechs(), updatedProgress,
                    newState.literacyRate(), newState.universities());
            } else {
                // Update progress
                newState = newState.withResearchProgress(tech.id(), newProgress);
            }
        }

        // Technology diffusion from trade
        newState = applyTechDiffusion(newState, techGraph, tradeConnectivity, random, events);

        return new ModuleResult<>(newState, events);
    }

    /**
     * Research points = base * (1 + literacyRate) * (1 + universityBonus)
     */
    private static double calculateResearchPoints(TechnologyState state, long population) {
        double literacyMultiplier = 1.0 + state.literacyRate();
        double universityMultiplier = 1.0 + (state.universities() * UNIVERSITY_BONUS);
        double populationFactor = Math.min(MAX_POPULATION_FACTOR, population / POPULATION_SCALING_DIVISOR);
        return BASE_RESEARCH_POINTS * literacyMultiplier * universityMultiplier * populationFactor;
    }

    /**
     * Find next technology that can be researched (all prerequisites met).
     */
    private static Optional<Technology> findNextResearchableTech(TechnologyState state, TechGraph graph) {
        return graph.getAllTechnologyIds().stream()
            .filter(id -> !state.unlockedTechs().contains(id))
            .filter(id -> isUnlocked(id, state.unlockedTechs(), graph))
            .map(graph::getTechnology)
            .min(Comparator.comparingDouble(Technology::researchCost));
    }

    /**
     * Check if all prerequisites are unlocked.
     */
    private static boolean isUnlocked(String techId, Set<String> unlocked, TechGraph graph) {
        return graph.getPrerequisites(techId).stream().allMatch(unlocked::contains);
    }

    /**
     * Apply technology diffusion based on trade connectivity.
     */
    private static TechnologyState applyTechDiffusion(
        TechnologyState state,
        TechGraph graph,
        double tradeConnectivity,
        SplittableRandom random,
        List<Event> events
    ) {
        // Simplified: Random chance to gain a tech from neighbors
        // In full implementation, would track neighbor civs
        return state;
    }
}

File: mobile/shared/src/commonMain/kotlin/org/flossware/civilization/module/TechnologyModule.kt
package org.flossware.civilization.module

import org.flossware.civilization.model.*
import kotlin.random.Random

object TechnologyModule {
    private const val BASE_RESEARCH_POINTS = 10.0
    private const val UNIVERSITY_BONUS = 0.5

    fun tick(
        current: TechnologyState,
        techGraph: TechGraph,
        tradeConnectivity: Double,
        population: Long,
        random: Random
    ): ModuleResult<TechnologyState> {
        val events = mutableListOf<Event>()
        var newState = current

        val researchPoints = calculateResearchPoints(current, population)
        val nextTech = findNextResearchableTech(current, techGraph)

        if (nextTech != null) {
            val currentProgress = current.researchProgress[nextTech.id] ?: 0.0
            val newProgress = currentProgress + researchPoints

            if (newProgress >= nextTech.researchCost) {
                newState = newState.withUnlockedTech(nextTech.id)
                events.add(Event(0, "", EventType.TECHNOLOGY_UNLOCKED, EventSeverity.MAJOR,
                    "Unlocked: ${nextTech.id}", nextTech.id))
                val updatedProgress = current.researchProgress.toMutableMap()
                updatedProgress.remove(nextTech.id)
                newState = TechnologyState(newState.unlockedTechs, updatedProgress, newState.literacyRate, newState.universities)
            } else {
                newState = newState.withResearchProgress(nextTech.id, newProgress)
            }
        }

        return ModuleResult(newState, events)
    }

    private fun calculateResearchPoints(state: TechnologyState, population: Long): Double {
        val literacyMultiplier = 1.0 + state.literacyRate
        val universityMultiplier = 1.0 + (state.universities * UNIVERSITY_BONUS)
        val populationFactor = minOf(1.0, population / 1_000_000.0)
        return BASE_RESEARCH_POINTS * literacyMultiplier * universityMultiplier * populationFactor
    }

    private fun findNextResearchableTech(state: TechnologyState, graph: TechGraph): Technology? {
        return graph.getAllTechnologyIds()
            .filter { it !in state.unlockedTechs }
            .filter { isUnlocked(it, state.unlockedTechs, graph) }
            .mapNotNull { graph.getTechnology(it) }
            .minByOrNull { it.researchCost }
    }

    private fun isUnlocked(techId: String, unlocked: Set<String>, graph: TechGraph): Boolean {
        return graph.getPrerequisites(techId).all { it in unlocked }
    }
}
